test/publish.test.js

```diff
@@ -19,25 +34,75 @@ test.afterEach.always(() => {
+
+  const result = await publish(pluginConfig, {env, options, nextRelease, logger: t.context.logger});
+
+  t.is(result.url, `https://gitlab.com/${encodedRepoId}/tags/${encodedGitTag}`);
+  t.deepEqual(t.context.log.args[0], ['Published GitLab release: %s', nextRelease.gitTag]);
+  t.true(gitlab.isDone());
+});
 
+test.serial('Publish a release with assets', async t => {
+  const cwd = 'test/fixtures/files';
+  const uploaded = {markdown: '[file.css](/uploads/file.css)', url: '/uploads/file.css', alt: 'file.css'};
+  const notes = `${nextRelease.notes}\n\n#### Assets\n\n* ${uploaded.markdown}`;
+  const assets = [['**', '!**/*.txt', '!.dotfile']];
   const gitlab = authenticate(env)
-    .post(`/projects/${encodedRepoId}/repository/tags/${encodedGitTag}/release`, {
-      tag_name: nextRelease.gitTag,
-      description: nextRelease.notes,
+    .post(releaseUrl, {...releaseBody, description: notes})
+    .reply(200);
+  const gitlabUpload = authenticate(env)
+    .post(`/projects/${encodedRepoId}/uploads`, body => {
+      return body.match('filename="file.css"');
     })
+    .reply(200, uploaded);
+
+  const result = await publish({assets}, {env, cwd, options, nextRelease, logger: t.context.logger});
+
+  t.is(result.url, `https://gitlab.com/${encodedRepoId}/tags/${encodedGitTag}`);
+  t.deepEqual(t.context.log.args[0], ['Uploaded file: %s', uploaded.url]);
+  t.deepEqual(t.context.log.args[1], ['Published GitLab release: %s', nextRelease.gitTag]);
+  t.true(gitlabUpload.isDone());
+  t.true(gitlab.isDone());
+});
+
+test.serial('Publish a release with array of missing assets', async t => {
+  const cwd = 'test/fixtures/files';
+  const emptyDirectory = tempy.directory();
+  const assets = [emptyDirectory, {path: 'missing.txt', label: 'missing.txt'}];
+  const gitlab = authenticate(env)
+    .post(releaseUrl, {...releaseBody, description: nextRelease.notes})
     .reply(200);
 
-  const result = await publish(pluginConfig, {env, options, nextRelease, logger: t.context.logger});
+  const result = await publish({assets}, {env, cwd, options, nextRelease, logger: t.context.logger});
 
   t.is(result.url, `https://gitlab.com/${encodedRepoId}/tags/${encodedGitTag}`);
   t.deepEqual(t.context.log.args[0], ['Published GitLab release: %s', nextRelease.gitTag]);
   t.true(gitlab.isDone());
 });
+
+test.serial('Publish a release with one asset and custom label', async t => {
+  const cwd = 'test/fixtures/files';
+  const uploaded = {markdown: '[file](/uploads/upload.txt)', url: '/uploads/upload.txt'};
+  const assetLabel = 'Custom Label';
+  const notes = `${nextRelease.notes}\n\n#### Assets\n\n* [${assetLabel}](${uploaded.url})`;
+  const assets = [{path: 'upload.txt', label: assetLabel}];
+  const gitlab = authenticate(env)
+    .post(releaseUrl, {...releaseBody, description: notes})
+    .reply(200);
+  const gitlabUpload = authenticate(env)
+    .post(`/projects/${encodedRepoId}/uploads`, body => {
+      return body.match('filename="upload.txt"');
```

Comment:
Similarly you can use a regexp for the assertion

Reply:
done